### PR TITLE
HPCC-28125 MP accept from client should use peer IP and not IP sent from client

### PR DESCRIPTION
**Pull request by mckellyln**
Signed-off-by: Mark Kelly <mark.kelly@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by mckellyln** [comment]
I am VERY concerned about this change.
But let me know your thoughts.

**Comment by mckellyln** [comment]
This all came about because Dali would log:
```
    void onClose(SocketEndpoint &ep)
    {
        StringBuffer clientStr;
        PROGLOG("Client closed (%s)", ep.getUrlStr(clientStr).str());
```
And the client IP was not correct, suspicious and causing alarm when a non-MP client (ssh) tried to connect.

**Comment by mckellyln** [comment]
The unittest crash is not related to this change, but it is from the change in https://track.hpccsystems.com/browse/HPCC-28102
I will submit a PR fix for that.

File: system/mp/mpcomm.cpp [hunk 1]
```diff
@@ -2237,6 +2237,37 @@ int CMPConnectThread::run()
...
+                    peerEp.getIpText(pIp);
+                    PROGLOG("MP: WARNING: remote client address (%s:%u) != peer address (%s:%u)", rIp.str(), _remoteep.port, pIp.str(), peerEp.port);
+                    _remoteep.set(peerEp);
+                    connectHdr.id[0].set(_remoteep);
+                    // should we sock->close() and continue ?
+                }
+
+                SocketEndpoint listenEp;
+                listensock->getEndpoint(listenEp);
+                unsigned short lPort = listenEp.port;
```
Comment (jakesmith) [inline system/mp/mpcomm.cpp:2257]: this is in order to get thte port of the listening socket..
rather than call getEndpoint which entails a system call etc., it would probably be better to store the port in the constructor as a member and reuse it here.

File: system/mp/mpcomm.cpp [hunk 2]
```diff
@@ -2237,6 +2237,37 @@ int CMPConnectThread::run()
...
+                IpAddress myLocalIp;
+                GetHostIp(myLocalIp);
```
Comment (jakesmith) [inline system/mp/mpcomm.cpp:2259]: this is going to be done per accept, can we move it to do once in CMPServer ? CMPConnectThread::run can access it via 'parent'

File: system/mp/mpcomm.cpp [hunk 3]
```diff
@@ -2237,6 +2237,37 @@ int CMPConnectThread::run()
...
+
+                if (!_remoteep.ipequals(peerEp))
+                {
+                    StringBuffer rIp, pIp;
+                    _remoteep.getIpText(rIp);
+                    peerEp.getIpText(pIp);
+                    PROGLOG("MP: WARNING: remote client address (%s:%u) != peer address (%s:%u)", rIp.str(), _remoteep.port, pIp.str(), peerEp.port);
+                    _remoteep.set(peerEp);
+                    connectHdr.id[0].set(_remoteep);
+                    // should we sock->close() and continue ?
```
Comment (jakesmith) [inline system/mp/mpcomm.cpp:2252]: Need to test/confirm/try to think of edge cases, but this change looks like the correct fix to me.
It is the peerEp that Dali should be using, not the what the client's notion of what its IP is

File: system/mp/mpcomm.cpp [hunk 4]
```diff
@@ -2237,6 +2237,37 @@ int CMPConnectThread::run()
...
+                }
+
+                SocketEndpoint listenEp;
+                listensock->getEndpoint(listenEp);
+                unsigned short lPort = listenEp.port;
+                IpAddress myLocalIp;
+                GetHostIp(myLocalIp);
+                if (!hostep.ipequals(myLocalIp))
```
Comment (jakesmith) [inline system/mp/mpcomm.cpp:2260]: this is going to be the norm. in the cloud I think (haven't confirmed), and not that unexpected in general.

I can see why the remoteep vs peerEp is a problem that will cause clashes etc.
That needs fixing and I think that fix looks correct.
But I'm not sure what problems the client's notion of what the Dali IP is, vs what Dali itself knows its IP is, is problematic?

File: system/mp/mpcomm.cpp [hunk 5]
```diff
@@ -2237,6 +2237,37 @@ int CMPConnectThread::run()
...
+                {
+                    StringBuffer hIp, lIp;
+                    hostep.getIpText(hIp);
+                    myLocalIp.getIpText(lIp);
+                    PROGLOG("MP: WARNING: remote client trying to reach (%s:%u) != listen address (%s:%u)", hIp.str(), hostep.port, lIp.str(), lPort);
+                    hostep.set(lPort, myLocalIp);
```
Comment (jakesmith) [inline system/mp/mpcomm.cpp:2266]: see comment above, I'm not sure this (client notion of Dali IP being different from Dali's notion of its IP) actually causes any problems...
but I am wondering if changing this (to Dali's notion of itself) may cause problems for the clients that will no longer see a match in packets coming back.

Probably pretty obvious if it breaks things on a cloud system where they will mismatch..

**Review by jakesmith — COMMENTED** [review]
@mckellyln - added some comments/questions

File: system/mp/mpcomm.cpp (hunk 3 above)
Comment (mckellyln) [reply system/mp/mpcomm.cpp:2252]: The concern I have is the port of the remoteep in a valid MP connection is really the other side's server port (how to connect to it), not the ephemeral port in the peerEp and so should we change it here ?

File: system/mp/mpcomm.cpp (hunk 1 above)
Comment (mckellyln) [reply system/mp/mpcomm.cpp:2257]: Yes, I can look into that.

File: system/mp/mpcomm.cpp (hunk 2 above)
Comment (mckellyln) [reply system/mp/mpcomm.cpp:2259]: Yes, good idea.

File: system/mp/mpcomm.cpp (hunk 5 above)
Comment (mckellyln) [reply system/mp/mpcomm.cpp:2268]: I agree, I am not sure changing hostep matters or causes problems.  Lets test and decide what is best.

**Comment by mckellyln** [comment]
Closing for now until we have some more information about MP clients and NAT
